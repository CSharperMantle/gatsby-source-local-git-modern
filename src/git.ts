import git from "simple-git/promise"
import { DefaultLogFields } from "simple-git/typings/response"

interface Author {
  id: string
  name: string
  email: string
}

interface Commit {
  id: string
  hash: string
  message: string
  body: string
  refs: string
  date: Date
  latest: boolean
  author: Author
}

interface Tag {
  id: string
  name: string
  latest: boolean
}

const createAuthor = (name: string, email: string): Author => ({
  name,
  email,
  id: `${name}__${email}`,
})

const getCommits = async (): Promise<Commit[]> => {
  const {
    all,
    latest: { hash: latestHash },
  } = await git().log<DefaultLogFields>()

  return all.map(({ hash, date, author_email, author_name, ...rest }) => ({
    hash,
    ...rest,
    id: hash,
    date: new Date(date),
    latest: hash === latestHash,
    author: createAuthor(author_name, author_email),
  }))
}

const getTags = async (): Promise<Tag[]> => {
  const { all, latest } = await git().tags()

  return all.map(name => ({
    name,
    id: name,
    latest: name === latest,
  }))
}

export { getCommits, getTags, Author, Commit, Tag }
