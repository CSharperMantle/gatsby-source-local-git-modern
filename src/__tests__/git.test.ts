import { getBranches, getCommits, getTags } from "../git"

test("First commit should match snapshot", async () => {
  const commits = await getCommits()
  expect(commits.pop()).toMatchSnapshot()
})

test("First tag should match snapshot", async () => {
  const [firstTag] = await getTags()
  expect(firstTag).toMatchSnapshot()
})

test("Exactly one branch should be current", async () => {
  const branches = await getBranches()
  expect(branches.filter(branch => branch.current)).toHaveLength(1)
})