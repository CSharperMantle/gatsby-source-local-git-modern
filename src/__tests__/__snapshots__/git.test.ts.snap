// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`First commit should match snapshot 1`] = `
Object {
  "author": Object {
    "email": "philipp.mudra@web.de",
    "id": "Philipp Mudra__philipp.mudra@web.de",
    "name": "Philipp Mudra",
  },
  "body": "",
  "date": 2019-11-10T18:07:04.000Z,
  "hash": "c003b915fa82ae86adcebd9f9461bbd215fd52eb",
  "id": "c003b915fa82ae86adcebd9f9461bbd215fd52eb",
  "latest": false,
  "message": "Initial commit",
  "refs": "",
}
`;

exports[`First tag should match snapshot 1`] = `
Object {
  "id": "v0.1.2",
  "latest": false,
  "name": "v0.1.2",
}
`;
